source/SpinalHDL/Libraries/Pipeline/introduction.rst

```diff
@@ -0,0 +1,762 @@
+
+.. image:: /asset/image/pipeline/simple_pip.png
+   :scale: 70 %
+
+Here is a simulation wave : 
+
+.. image:: /asset/image/pipeline/simple_wave.png
+
+Here is the same example but using more of the API :
+
+
+.. code-block:: scala
+
+    import spinal.core._
+    import spinal.core.sim._
+    import spinal.lib._
+    import spinal.lib.misc.pipeline._
+
+    class TopLevel extends Component {
+      val VALUE = Stageable(UInt(16 bits))
+
+      val io = new Bundle{
+        val up = slave Stream(VALUE)  //VALUE can also be used as a HardType
+        val down = master Stream(VALUE)
+      }
+      
+      // NodesBuilder will be used to register all the nodes created, connect them via stages and generate the hardware
+      val builder = new NodesBuilder()
+
+      // Let's define a Node which connect from io.up
+      val n0 = new builder.Node{
+        arbitrateFrom(io.up)
+        VALUE := io.up.payload
+      }
+
+      // Let's define a Node which do some processing
+      val n1 = new builder.Node{
+        val RESULT = insert(VALUE + 0x1200)
+      }
+
+      //  Let's define a Node which connect to io.down
+      val n2 = new builder.Node {
+        arbitrateTo(io.down)
+        io.down.payload := n1.RESULT
+      }
+
+      // Let's connect those nodes by using registers stages and generate the related hardware
+      builder.genStagedPipeline()
+    }
+
+Stageable
+============
+
+Stageable class can be instanciated to represent some data which can go through the pipeline. Technicaly speaking, Stageable is a HardType which has a name and is used as a "key" to retrieve stuff.
+
+.. code-block:: scala
+    
+    val PC = Stageable(UInt(32 bits))
+    val PC_PLUS_4 = Stageable(UInt(32 bits))
+
+    val n0, n1 = Node()
+    val s01 = StageConnector(n0, n1)
+
+    n0(PC) := 0x42
+    n1(PC_PLUS_4) := n1(PC) + 4
+
+Note that I got used to name the Stageable instances using uppercase. This is to make it very explicit that the thing isn't a hardware signal, but are more like a "key" to access things.
+
+Node
+============
+
+Node mostly host the valid/ready arbitration signal, and the hardware signal required for all the Stageable values going through it.
+
+You can access its arbitration via :
+
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 5
+
+   * - API
+     - Description
+   * - node.valid
+     - Is the signal which specify if a transaction is present on the node
+   * - node.ready
+     - Is the signal which specify if the node transaction can move away.
+   * - node.isValid
+     - node.valid's read only accessor
+   * - node.isReady
+     - node.ready's read only accessor
+   * - node.isFiring
+     - True when the node transaction is successfuly moving futher (isValid && isReady && !isRemoved). Usefull to commit state changes
+   * - node.isMoving
+     - True when the node transaction is moving (isValid && (isReady || isRemoved)). Usefull to "reset" states
+   * - node.isRemoved
+     - True when the node is being flushed
```

Comment:
This is visible this cycle, when this stage has been earmarked to be flushed at the end of the current cycle.  At the next advancement of data (for example at the next cycle) this will cause the signal data in this stage to be discarded and the isValid for the next stage to be made inactive.  Thus deactivating/discarding the data present in this stage at this time.